keywrap/pkcs11/keywrapper_pkcs11.go

```diff
@@ -0,0 +1,204 @@
+	return nil, nil
+}
+
+// GetRecipients converts the wrappedKeys to an array of recipients
+// We cannot do this with pkcs11
+func (kw pkcs11KeyWrapper) GetRecipients(packet string) ([]string, error) {
+	return []string{"[pkcs11]"}, nil
+}
+
+// getRSA generate a rsa key if it doesn't exist
+func getRSA(label string, p *pkcs11.Ctx, sh pkcs11.SessionHandle) (pub, priv pkcs11.ObjectHandle) {
+	pub, err := findObject(p, sh, pkcs11.CKO_PUBLIC_KEY, label)
+	if err != nil {
+		pub, priv = generateRSAKeyPair(p, sh, label, true)
+	}
+	return
+}
+
+func findObject(p *pkcs11.Ctx, sh pkcs11.SessionHandle, class uint, label string) (pkcs11.ObjectHandle, error) {
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, class),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+	}
+	if err := p.FindObjectsInit(sh, template); err != nil {
+		return 0, errors.Wrap(err, "FindObjectsInit")
+	}
+	obj, _, err := p.FindObjects(sh, 1)
+	if err != nil {
+		return 0, errors.Wrap(err, "FindObjects")
+	}
+	if err := p.FindObjectsFinal(sh); err != nil {
+		return 0, errors.Wrap(err, "FindObjectsFinal")
+	}
+	if len(obj) > 0 {
+		return obj[0], nil
+	}
+	return 0, errors.New("Not Found Object")
+}
+
+func generateRSAKeyPair(p *pkcs11.Ctx, session pkcs11.SessionHandle, tokenLabel string, tokenPersistent bool) (pkcs11.ObjectHandle, pkcs11.ObjectHandle) {
+	publicKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_RSA),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, tokenPersistent),
+		pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, []byte{1, 0, 1}),
+		pkcs11.NewAttribute(pkcs11.CKA_MODULUS_BITS, 2048),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, tokenLabel),
+	}
+	privateKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, tokenPersistent),
+		pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, tokenLabel),
+		pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+		pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, true),
+	}
+	pbk, pvk, e := p.GenerateKeyPair(session,
+		[]*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS_KEY_PAIR_GEN, nil)},
+		publicKeyTemplate, privateKeyTemplate)
+	if e != nil {
+		fmt.Printf("failed to generate keypair: %s\n", e.Error())
+	}
+
+	return pbk, pvk
+}
+
+// loginDevice login Device and use the slot[0]
+func loginDevice(modules [][]byte, pins [][]byte) (ctx *pkcs11.Ctx, session pkcs11.SessionHandle, err error) {
+	if len(pins) == 0 {
+		return nil, 0, errors.New("Need input module pin")
+	}
+	pin := string(pins[0])
+	pin = strings.TrimSpace(pin)
+
+	if len(modules) > 1 {
+		return nil, 0, errors.New("Just support *one* module")
+	}
+	module := string(modules[0])
+	module = strings.TrimSpace(module)
+	if module == "" {
+		return nil, 0, errors.New("Please check Module path, input is: " + module)
+	}
+
+	ctx = pkcs11.New(module)
```

Comment:
can ctx be null? may want to check for it.

Reply:
fixed this.